src/storage/cmemlog.c

```diff
@@ -0,0 +1,207 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <stdlib.h>
+#include <stdio.h>
+#include "cmemlog.h"
+
+
+/**
+ * Initialize the memlog for each sqlite db connection
+ * mem_log - pointer to the memlog
+ * initial_size - the initial array size of the store 
+ * */
+Status hustle_memlog_initialize(HustleMemLog **mem_log, int initial_size) {
+  if (initial_size <= 0) {
+      return MEMLOG_ERROR;
+  }
+  *mem_log = (HustleMemLog*)malloc(sizeof(HustleMemLog)); 
+  (*mem_log)->record_list = (DBRecordList *)malloc(initial_size * sizeof(DBRecordList));
+  (*mem_log)->total_size = initial_size;
+  int table_index = 0;
+  while (table_index < (*mem_log)->total_size) {
+    (*mem_log)->record_list[table_index].head = NULL;
+    (*mem_log)->record_list[table_index].tail = NULL;
+    (*mem_log)->record_list[table_index].curr_size = 0;
+    table_index++;
+  }
+  return MEMLOG_OK;
+}
+
+/**
+ * Create a DBRecord - each node in the linkedlist.
+ * data - SQLite's data record format with header in the begining
+ * nData - the size of the data
+ * */
+DBRecord* hustle_memlog_create_record(const void *data, int nData) {
+    if (data == NULL) {
+      return NULL;
+    }
+    DBRecord* record = (DBRecord*) malloc(sizeof(DBRecord));
+    record->data = data;
+    record->nData = nData;
+    record->next_record = NULL;
+    return record;
+}
+
+/**
+ * Insert's the record to the memlog and grows the array size, if the table id
+ * is greater than the array size.
+ * 
+ * mem_log - pointer to the memlog
+ * record - DBRecord needs to be inserted
+ * table_id - root page id of the table
+ * */
+Status hustle_memlog_insert_record(HustleMemLog *mem_log, DBRecord *record, int table_id) {
+  if (mem_log == NULL || record == NULL) {
+    return MEMLOG_ERROR;
+  }
+  if (table_id >= mem_log->total_size) {
+    int old_table_list_size = mem_log->total_size;
+    mem_log->total_size *= 2;
```

Comment:
Can the table_id be > the new mem_log->total_size? then line 87 would access unallocated memory